images/pre-puller/README.md

```diff
@@ -0,0 +1,42 @@
+# Image Pre-Puller
+
+## What?
+A small go program that does the following:
+
+1. Check if the given list of images already exist on all scheduleable nodes.
+   If they do, exit!
+2. If they don't, create a deamonset (the spec for which is passed in
+   as the `daemonset-spec` commandline parameter)
+3. Check every 2s if the images are present in all scheduleable nodes.
+4. Once image is in all schedulable nodes, kill the daemonset created in (2) and exit
+
+## Why?
+
+This is used as a [helm hook](https://github.com/kubernetes/helm/blob/master/docs/charts_hooks.md)
+to wait for the user images to be present on all nodes before we restart the
+hub. This cuts down the amount of time it takes for a user server to start,
+since the image no longer needs to be pulled. For large images this can cut down
+startup time from almost ten minutes to a few seconds.
+
+## Why Go?
+
+The role of this program is to run as a pod whenever the user is running `helm
+install` or `helm upgrade` and help pull in larger images. The image that this
+program requires to run must be as small as possible - it is unideal to pull in
+a multi hundred megabyte image first to check if more images need to be pulled
+in. This is particularly bad in the `helm upgrade` case when new images to *not*
+need to be pulled in - then we end up pulling in a multi-hundred mb image that
+then just exits pretty much immediately.
+
+This gives us a few choices:
+
+1. Bash
+2. Go
+3. Rust
+
+All these can produce images < 10MB. Originally this was in Bash, but it very
+quickly became hard to maintain (since we need to set up TLS and talk to a JSON
+HTTP API & munge the resulting JSON). There was very little error handling code,
+and extending it was very bad. This made it clear it needed to be in a real
+programming language. Go is a little bit more 'mature' now than Rust is, and so
```

Comment:
Just a thought here - there are a few moments where you (or the docs) give opinionated statements about the relative merits of Bash / Go / etc. Do you know of at least a blog post we could link people to that discusses any of this? I believe you that Go is a better language to implement this in, but I could see, e.g., a person who had recently taken software carpentry (which teaches bash and tells students that they should *really* learn bash in doing their analyses) getting frustrated by the opinionated language w/o any material to back it up

Reply:
I dont' think that in any way this diminishes the value of Bash or Rust, I think it does just show that different times in the lifetime of a project might require different languages. Bash was nice at the beginning (and is ubiquitous), now we just need a more robust language. 

There is of course the fact that the language you know is likely better suited as you know it.

Reply:
I don't think there's any shame in saying "Hey, this particular component outgrew bash.  Once you find yourself really wanting a map data type, it's probably time to consider whether sticking to shell is the right thing."

IMHO Rust is about 4-5 years behind Go in terms of maturity of its toolset and ecosystem.  I also find its object lifecycle management harder to wrap my head around than Go, but I cut my teeth on C and so I have a bias towards Go, since in many ways it really is C with 30 years of lessons learned.

Reply:
I totally agree with you - I'm just saying we need to be careful about saying "X is better than Y" without providing evidence for it. It's important to be careful not to accidentally introduce the kind of "my tools are better than yours" mentality that can sometimes have unintended consequences on the reader, especially those who are relatively inexperienced in the devops world (e.g. all scientists). E.g., linking out to a page like [this](http://goroutines.com/shell) or [this](https://eng.heetch.com/go-as-a-scripting-language-eeaa0da842ea) could help.

Reply:
I wrote out a really long reply here but then realized I was arguing about relative benefits of programming languages, which I've a personal ban on :)

I've instead pushed a commit that removes all references to bash or rust from the README. I wrote it that way to have it pre-empt any language wars, but now it's clear that it'll have exactly the opposite effect. LMK if this works for you, @choldgraf?

Also, I appreciate you thinking about how this might feel incendiary to a population that I do not have much insights to. I agree with what everyone else has said in this thread, but also agree that we should take extra steps to avoid putting anything that might be even vaguely interpreted as 'X sucks' in an official, non-personal capacity. Thanks!

Reply:
The only sentiment I'm trying to convey here is:

1. Why is this in Go while the rest of the project is in Python?
2. Because we want to have a very small image size.

The fact that this code was in bash at some point is a historical accident that seems best to ignore in our docs.

Reply:
Ok, I simplified it even more, making it into a FAQ that addresses both @choldgraf's concerns here and @cam72cam's questions about why we aren't using the kubernetes go client library.

Reply:
I've also edited this PR title and description to remove things that could possibly be used as ammo in any 'X sucks' conversation.